Salt reconfiguration to get rid of nginx on GCE

roberthbailey opened the pull request:
Instead, pass a bearer token for kubectl to the apiserver.
- install libcap2-bin w/ apiserver
- use setcap to allow the apiserver to bind to low numbered ports
- set secure port for the apiserver to 443
- Kubelet talk to port 443 on gce

roberthbailey commented:
I've tested that cluster creation works (as does kubectl for the new cluster) but haven't tested cluster upgrade or run e2e tests yet.

roberthbailey commented:
@smarterclayton @deads2k @liggitt - Hopefully this is more along the lines of what you were looking for than #6508.

lavalamp commented:
heavy_breathing_cat.gif

zmerlynn commented:
Surfacing briefly.. I'm still OOO and probably won't get to this until
tomorrow or so. Access and time has been spottier than anticipated.
On Apr 9, 2015 1:37 PM, "Daniel Smith" <notifications@github.com> wrote:

> heavy_breathing_cat.gif
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6618#issuecomment-91303454>
> .
>

roberthbailey commented:
@zmerlynn - don't worry. I wanted to get some early feedback from the RedHat folks on this. I don't want it merged before the 0.15 cut next week.

deads2k commented on cluster/common.sh:
```diff
@@ -101,3 +109,27 @@ function get-kubeconfig-basicauth() {
+  # is missing.
+  # Note: we save dot ('.') to $root because the 'with' action overrides it.
+  # See http://golang.org/pkg/text/template/.
+  local token='{{$root := .}}{{with index $root "current-context"}}{{with index $root "contexts" .}}{{with index . "user"}}{{with index $root "users" .}}{{index . "token"}}{{end}}{{end}}{{end}}{{end}}'
```
This format just changed (today https://github.com/GoogleCloudPlatform/kubernetes/pull/6476).  See https://gist.github.com/deads2k/c294a882c9358ea7679f for examples of the new template format.

roberthbailey replied:
Thanks for the heads up @deads2k. I'm waiting on a few things before rebasing this:

1. We need to do a bit of work in GKE to make this a smooth transition. It'll take a couple of weeks for us to get things in place.

2. Once the apiserver runs in a container, then I won't need to install the libcap2-bin package or run /sbin/setcap on the apiserver, which will simplify things a bit.

roberthbailey replied:
I've fixed this to use the new template format.

deads2k commented on cluster/common.sh:
```diff
@@ -44,11 +45,18 @@ function create-kubeconfig() {
-                                                --client-certificate="${KUBE_CERT}" \
-                                                --client-key="${KUBE_KEY}" \
-                                                --embed-certs=true
+  if [[ -z "${KUBE_USER:-}" || -z "${KUBE_PASSWORD:-}" ]]; then
```
Also check to see if a bearer token is present?

roberthbailey replied:
What would you recommend doing if it isn't? The previous code just assumes that all of the necessary vars are set, and after a transition period I'd like to remove the username/password option from all cloud providers and transition everyone to using bearer tokens. I suppose we could skip writing any credentials to the kubeconfig file if they aren't set, but that means there is a bug in the startup scripts and the user won't be able to interact with their cluster.

erictune commented:
This looks awesome.

zmerlynn commented on cluster/gce/configure-vm.sh:
```diff
@@ -245,23 +245,17 @@ EOF
     (umask 077;
-      echo "${KUBELET_TOKEN},kubelet,kubelet" > "${KNOWN_TOKENS_FILE}")
+      echo "${KUBE_BEARER_TOKEN},admin,admin" > "${KNOWN_TOKENS_FILE}";
+      echo "${KUBELET_TOKEN},kubelet,kubelet" >> "${KNOWN_TOKENS_FILE}")
```
If I'm being pedantic, this has to handle the upgrade case. See the note on line 273/267 below.

zmerlynn replied:
Maybe this should actually be split up into a token directory and reassembled here? (Like how .conf files are done with foo.d/* directories?). Just a random thought so that any future upgrades are less obnoxious.

roberthbailey replied:
I think your suggestions are orthogonal to this PR (seeing as how the token file already has many files with the service account tokens) but they are good points. Since we don't yet support upgrade, I'm leaning towards leaving this as a breaking change between 0.15.0 and 0.16.0.

zmerlynn replied:
SGTM

zmerlynn commented on cluster/saltbase/salt/kube-apiserver/default:
```diff
@@ -44,6 +44,12 @@
 {% set token_auth_file = "--token_auth_file=/dev/null" -%}
 
 {% if grains.cloud is defined -%}
+{% if grains.cloud in [ 'gce' ] -%}
```
Optimism? Funny form of `==` :)

(But I agree that it'll probably be more than one eventually.)

roberthbailey replied:
Optimism. I can make it a straight == if you'd prefer until we have other cloud providers working.

zmerlynn commented on cluster/saltbase/salt/kube-apiserver/default:
```diff
@@ -44,6 +44,12 @@
 {% set token_auth_file = "--token_auth_file=/dev/null" -%}
 
 {% if grains.cloud is defined -%}
```
I believe these can be compounded as `grains.cloud is defined and `

zmerlynn commented on cluster/saltbase/salt/kube-apiserver/initd:
```diff
@@ -41,6 +41,9 @@ do_start()
         ulimit -n 65536
 
+        # Make it possible to bind to low numbered ports.
+        /sbin/setcap 'cap_net_bind_service=+ep' $DAEMON
```
The `libpcap` install is conditional but the `setcap` call isn't?

roberthbailey replied:
This was temporary until kube-apiserver was running in a pod. Now that it's in a pod, I think I can get rid of this call and the extra deb package. Will rebase and do further testing.

roberthbailey replied:
Now the the apiserver is running in a container, this has cleaned up quite nicely.

Commit:
Remove nginx and replace basic auth with bearer token auth for GCE.

 - Configure the apiserver to listen securely on 443 instead of 6443.
 - Configure the kubelet to connect to 443 instead of 6443.
 - Update documentation to refer to bearer tokens instead of basic auth.

roberthbailey commented:
Rebased. PTAL.

zmerlynn commented:
LGTM (sorry, rebase again)

roberthbailey commented:
Rebased. I also updated some documentation files to talk about bearer tokens instead of basic auth. /cc @jlowdermilk

roberthbailey commented:
Two issues I'm working on.

1. The create external load balancer e2e test is failing due to a timeout:
```
Summarizing 1 Failure:

[Fail] Services [It] should be able to create a functioning external load balancer 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:339

Ran 35 of 40 Specs in 1244.857 seconds
FAIL! -- 34 Passed | 1 Failed | 1 Pending | 4 Skipped F0420 22:45:12.452252   99306 driver.go:94] At least one test failed
```

2. `go run hack/e2e.go -v --up` is failing to validate the e2e cluster
```
Running: /Users/robertbailey/Work/go/src/github.com/roberthbailey/kubernetes/hack/e2e-internal/../../cluster/../cluster/gce/../../cluster/../cluster/../cluster/gce/../../cluster/../cluster/../cluster/gce/../../cluster/../_output/dockerized/bin/darwin/amd64/kubectl get cs
2015/04/20 23:05:23 Error running up: exit status 1
2015/04/20 23:05:23 Error starting e2e cluster. Aborting.
exit status 1
```

roberthbailey commented:
`go run hack/e2e.go --up` is failing because of #7139 (and it's broken for me on the master branch as well).

roberthbailey commented:
@a-robinson pointed out that the reason the e2e test was failing was that I didn't have a firewall to allow http requests in my project. It's now passing as well:

```
$ go run hack/e2e.go -v --test --test_args="--ginkgo.focus=Service.*functioning external load balancer.*"
...
Ran 1 of 40 Specs in 61.238 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 39 Skipped I0421 15:34:21.475133   42739 driver.go:96] All tests pass
```

roberthbailey commented:
Now that the latest gcloud release is out, we can move forward with this PR. I rebased on master and re-ran e2e tests. services.sh failed (with the same error I'm seeing in the latest run in Jenkins) but everything else passed.

zmerlynn commented:
LGTM.

![nginx-tombstone](https://cloud.githubusercontent.com/assets/4942464/7283014/74e6ec6e-e8e8-11e4-9e5f-de57bb07685f.jpg)

zmerlynn commented:
Waiting on green, and feedback to https://groups.google.com/forum/#!topic/kubernetes-dev/nf9L-IeB56k
